tough/src/transport.rs

```diff
@@ -1,36 +1,172 @@
-use std::io::Read;
+#[cfg(feature = "http")]
+use crate::{ClientSettings, HttpTransport};
+use dyn_clone::DynClone;
+use snafu::Snafu;
+use std::fmt::Debug;
+use std::io::{ErrorKind, Read};
 use url::Url;
 
 /// A trait to abstract over the method/protocol by which files are obtained.
-pub trait Transport {
-    /// The type of `Read` object that the `fetch` function will return.
-    type Stream: Read;
+///
+/// The trait hides the underlying types involved by returning the `Read` object as a
+/// `Box<dyn Read + Send>` and by requiring concrete type [`TransportError`] as the error type.
+///
+pub trait Transport: Debug + DynClone {
+    /// Opens a `Read` object for the file specified by `url`.
+    fn fetch(&self, url: Url) -> Result<Box<dyn Read + Send>, TransportError>;
+}
 
-    /// The type of error that the `fetch` function will return.
-    type Error: std::error::Error + Send + Sync + 'static;
+// Implement `Clone` for `Transport` trait objects.
+dyn_clone::clone_trait_object!(Transport);
```

Comment:
If I'm reading `dyn_clone`'s documentation correctly, any implementations of Transport just need to implement `Clone` to meet this constraint, right? Can we make that clearer in the documentation?